### ISSUE TITLE
Conflictual type union function/object

### ISSUE THREAD

**Issue by baanloh** [microsoft/TypeScript#39310]
**TypeScript Version:**  4.0.0-dev.20200627

**Search Terms:** type union function object

**Code**

```ts
type Callback = () => string | {
    A: () => string;
    B: () => string;
};

function Foo(e: Callback) {
    if (typeof e === "function") {
        console.log(e());
    } else {
        // here e is never
        const a = e.A();
        const b = e.B();
        console.log({a, b});
    }
}

Foo(() => "salut");
Foo({
    A: () => "A",
    B: () => "B",
});
// Argument of type '{ A: () => string; B: () => string; }' is not assignable to parameter of type 'Callback'.
// Object literal may only specify known properties, and 'A' does not exist in type 'Callback'.
```

**Expected behavior:** 
The second call to `Foo` must not raise an error and the type guard `typeof e === "function"` must separate the object and the function.

**Actual behavior:**
The second call of `Foo` make an error.
The typeguard `typeof e === "function"` not work properly.

**Playground Link:** [link](https://www.typescriptlang.org/play?ts=4.0.0-dev.20200627#code/C4TwDgpgBAwghgGwQIzgYwNZQLxQBQCUOAfFAM7ABOAlgHYDmUAPlAN4BQUXUAggFz4i2UhRoMA3J24AhAYRLkqdepIC+k9gDMArrTTBqAe1pQAYocN4IA+ElSYiHblGqb8oSIbfRsvqACIdPQNjf0cpZy40YzJDBAgAOgRDeitCAklnVSgIBDJoJ0iuAHpiqAALCEpoaGoyKFoIADcqiMjo2gooOBwchJ5CTKKoDq7kXsTpQbbnUbjE5NTWOAAaKGRVDLbVdh32c0t5YQCyRG1gMMkDvEKufkEFfx5-FbbZB+P-aRfdrdLeSj0bQAWwgtGAUC8UA80AA5KxeHIhCIlBIoO8jiixCooKpYS56rRDBC4GQyNR6LQ4Mh4tDDFAwHBKHBQcAqpC3DCoLDbCh0BhYQl2P8APLIABWEH0UAQ1DZzIQUGBcBAkNoCFVZEgaFcqowRIA7iYwJRDJBKAYIGQ1nBaAATbk8fF2wxWhrEnIADzqELo0PAcN59gFCSAA)

**Comment by j-oliveras** [comment]
You declared `Callback` as a function that returns an string or an object with properties `A` and `B`: it is interpreted as:
```ts
type Callback = () => (string | {
    A: () => string;
    B: () => string;
});
```

Add parenthesis around function and it will work:
```ts
type Callback = (() => string) | {
    A: () => string;
    B: () => string;
};
```
